### PR TITLE
Configurable tests

### PR DESCRIPTION
**Pull request by natemcmaster**
Make SqlServer testing configurable. 
~~Add makefile target for running a test matrix.~~

Required to test against more database platforms  #2897 and #2898.
Also provides a way to automate these tests.

### CODE REVIEW

File: makefile.shade [hunk 1]
```diff
@@ -105,6 +107,80 @@ var sqliteTestDir='test/EntityFramework.Sqlite.FunctionalTests/bin/${E("Configur
...
+        foreach(var test in tests.Keys)
+        {
+            Log.Info(string.Format("Running test '{0}'", test));
+            var testSettings = GetSection(tests, test);
+
+            var cmd = GetValue(testSettings, "cmd") as string;
+
+            var kCmd = GetValue(testSettings, "dnx") as string;
+            var folder = GetValue(testSettings, "folder") as string;
+            var dnvmUse = GetValue(testSettings,"dnvmUse") as string;
```
Comment (smitpatel) [inline makefile.shade]: space after ,

File: makefile.shade [hunk 2]
```diff
@@ -105,6 +107,80 @@ var sqliteTestDir='test/EntityFramework.Sqlite.FunctionalTests/bin/${E("Configur
         }
     }
 
+var matrixTestConfig='matrixTests.json'
+
+#test-matrix
```
Comment (bricelam) [inline makefile.shade]: We can probably capture this in CI configurations instead of the makefile.

**Comment by natemcmaster** [comment]
Ping!

**Comment by bricelam** [comment]
Looks fine to me. You should get input from a few others before merging.

**Comment by natemcmaster** [comment]
@rowanmiller this is most similar to what you did with benchmarks config. Thoughts?

File: test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestConfig.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        }
+
+        public static SqlServerTestConfig Default()
+        {
+            var config = new ConfigurationBuilder(".")
+                .AddJsonFile("config.json", optional: true)
+                .AddJsonFile("config.test.json", optional: true)
+                .AddEnvironmentVariables()
+                .Build();
+
```
Comment (rowanmiller) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestConfig.cs]: Should we be doing some caching here? In the benchmarks I had a static Instance property so that we only ever ran all this logic once.

File: test/EntityFramework.SqlServer.FunctionalTests/README.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
+EntityFramework.SqlServer.FunctionalTests
```
Comment (rowanmiller) [inline test/EntityFramework.SqlServer.FunctionalTests/README.md]: I think it's fine to leave this in... but I'm not sure how much value it adds. I think it's pretty obvious how to use the config file and this will probably just get out of date.

**Comment by rowanmiller** [comment]
:shipit: from my perspective. Maybe others will have pushback on the static-ness of the config... but I think it makes sense.

File: test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestConfig.cs (hunk 3 above)
Comment (natemcmaster) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestConfig.cs]: I don't cache because some tests modify the ConnectionStringBuilder to add custom properties.

File: test/EntityFramework.SqlServer.FunctionalTests/README.md (hunk 4 above)
Comment (natemcmaster) [inline test/EntityFramework.SqlServer.FunctionalTests/README.md]: K good. Then i'm gonna remove to reduce update-burden.

File: test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestConfig.cs (hunk 3 above)
Comment (rowanmiller) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestConfig.cs]: I wonder if we should cache the base one and then have With return a new instance of the builder

Comment (rowanmiller) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestConfig.cs]: I'm only thinking of perf though... if it is fast enough then it doesn't matter :smile:

Comment (natemcmaster) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestConfig.cs]: The slowest part is probably reading config from file. I can easily cache that part.

File: test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestConfig.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            int connectTimeout;
+            if (!int.TryParse(config[prefix + "ConnectTimeout"], out connectTimeout))
```
Comment (divega) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestConfig.cs]: Nit: Can you make this a bit cleaner by store the configuration section for the prefix in a variable? E.g.:
```C#
var section = config.GetSection(prefix);
// ...
if (!int.TryParse(section["ConnectTimeout"], out connectTimeout)) 
```

**Comment by natemcmaster** [comment]
Responded to feedback from @divega  and @rowanmiller

File: test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestConfig.cs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+            return this;
+        }
+
+        public SqlServerTestConfig With(Action<SqlConnectionStringBuilder> connStrBuilder)
+        {
+            connStrBuilder(ConnectionStringBuilder);
+            return this;
+        }
+
+        public SqlConnectionStringBuilder ConnectionStringBuilder
```
Comment (divega) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestConfig.cs]: Is anyone using this? Maybe make pvt?

File: test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestConfig.cs (hunk 3 above)
Comment (divega) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestConfig.cs]: > I don't cache because some tests modify the ConnectionStringBuilder to add custom properties.

But aren't you deferring the creation of the ConnectionStringBuilder until later?

File: test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestConfig.cs [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+            var testConfig = new SqlServerTestConfig
+            {
+                DataSource = config["DataSource"] ?? @"(localdb)\MSSQLLocalDB",
+                IntegratedSecurity = integratedSecurity,
+                ConnectTimeout = connectTimeout,
+                UserID = config["UserID"],
+                Password = config["Password"]
+            };
+            Debug.WriteLine($"Using '{testConfig.ConnectionString}'");
+            return testConfig;
```
Comment (divega) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestConfig.cs]: Still think @rowanmiller is right and you can cache this because anyway you haven't created the string builder that others will modify. Of course, if this wasn't test code, I would also recommend making it immutable :smile:

Comment (divega) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestConfig.cs]: Oh, it already is immutble, isn't it?

Comment (divega) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestConfig.cs]: Ah, I think I see what the difficulty is: the connection string builder you keep around is mutable and tests can change that. I would recommend removing the field and creating the connection string builder every time. It would require removing the version of With() that takes a key and an object (it is redundant anyway) and changing the other With() to return the connection string directly (and maybe rename it?)

File: test/EntityFramework.SqlServer.FunctionalTests/MigrationsSqlServerFixture.cs [hunk 8]
```diff
@@ -16,13 +15,9 @@ protected override void ConfigureServices(IServiceCollection services)
...
         {
-            optionsBuilder.UseSqlServer("" +
-                new SqlConnectionStringBuilder
-                {
-                    DataSource = @"(localdb)\MSSQLLocalDB",
-                    InitialCatalog = nameof(MigrationsSqlServerTest),
-                    IntegratedSecurity = true
-                });
+            optionsBuilder.UseSqlServer(SqlServerTestConfig.Default()
+                .With(b=> { b.InitialCatalog = nameof(MigrationsSqlServerTest); })
```
Comment (anpete) [inline test/EntityFramework.SqlServer.FunctionalTests/MigrationsSqlServerFixture.cs]: With what? Assuming this is connection string config? NIT: Formatting and braces not required.

File: test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestConfig.cs [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        {
+            connStrBuilder(ConnectionStringBuilder);
+            return this;
+        }
+
+        public SqlConnectionStringBuilder ConnectionStringBuilder
+        {
+            get
+            {
+                if (_connectionStringBuilder != null)
```
Comment (anpete) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestConfig.cs]: Use blank lines between blocks and members.

File: test/EntityFramework.SqlServer.FunctionalTests/MigrationsSqlServerFixture.cs (hunk 8 above)
Comment (anpete) [inline test/EntityFramework.SqlServer.FunctionalTests/MigrationsSqlServerFixture.cs]: Can this test be standardized away from OnConfiguring?

File: test/EntityFramework.SqlServer.FunctionalTests/MonsterFixupSqlServerTest.cs [hunk 10]
```diff
@@ -45,16 +44,13 @@ protected override DbContextOptions CreateOptions(string databaseName)
...
-            return new SqlConnectionStringBuilder
-            {
-                DataSource = @"(localdb)\MSSQLLocalDB",
-                MultipleActiveResultSets = true,
-                InitialCatalog = name,
-                IntegratedSecurity = true,
-                ConnectTimeout = 30
-            }.ConnectionString;
-        }
+            => SqlServerTestConfig.Default()
```
Comment (anpete) [inline test/EntityFramework.SqlServer.FunctionalTests/MonsterFixupSqlServerTest.cs]: Can this test follow the standard fixture pattern?

File: test/EntityFramework.SqlServer.FunctionalTests/NavigationTest.cs [hunk 11]
```diff
@@ -90,7 +90,15 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
 
         protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
         {
-            optionsBuilder.UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=StateManagerBug;Trusted_Connection=True;MultipleActiveResultSets=true");
+            var config = SqlServerTestConfig.Default()
```
Comment (anpete) [inline test/EntityFramework.SqlServer.FunctionalTests/NavigationTest.cs]: Can this test follow the standard fixture pattern?

File: test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestConfig.cs [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Data.SqlClient;
+using Microsoft.Framework.Configuration;
+using System.Diagnostics;
+using System;
+
+namespace Microsoft.Data.Entity.SqlServer.FunctionalTests
+{
+    public class SqlServerTestConfig
```
Comment (divega) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestConfig.cs]: Half of this class looks like a DTO for a connection string builder, and I am not sure that is needed. E.g. we could just use connection string builders, an extension method to populate it from configuration, and a default populated instance we cache (and we promise tests will never change :smile:). But give that this is test code and that there can be different opinions I don't care too much.

**Comment by natemcmaster** [comment]
@divega @anpete I will revisit my sloppy implementation of `SqlServerTestConfig`. I didn't give it much design as this is test code. But I'll improve it so test writers understand better how to use it. I'll update the PR when its ready.

**Comment by natemcmaster** [comment]
Updated. Moved loading environment variables into an extension method on `SqlConnectionStringBuilder`

File: test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlConnectionStringBuilderExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+            if (overwrite || string.IsNullOrEmpty(builder.Password))
+            {
+                builder.Password = config["Password"];
+            }
+
+            return builder;
+        }
+
+        private static readonly Lazy<IConfiguration> _sqlTestConfiguration = new Lazy<IConfiguration>(() =>
```
Comment (divega) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlConnectionStringBuilderExtensions.cs]: Is `Lazy<T>` necessary?

File: test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestStore.cs [hunk 14]
```diff
@@ -366,7 +366,7 @@ private SqlServerTestStore CreateTransient(bool createDatabase)
 
                     await command.ExecuteNonQueryAsync();
 
-                    var userFolder = Environment.GetEnvironmentVariable("USERPROFILE");
+                    var userFolder = Environment.GetEnvironmentVariable("USERPROFILE") ?? Environment.GetEnvironmentVariable("HOME");
```
Comment (divega) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestStore.cs]: Interesting to see this code... Maybe there is a good reason we are doing it in this case, but simply deleting the files has traditionally been a broken way of dropping a SQL Server database, even on LocalDB. An usual consequence AFAIR is that next time you try to create it you'll get a failure and you will have to try opening it and closing it before SQL Server records the fact that it doesn't exist anymore.

Comment (divega) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestStore.cs]: Ah, I just saw the drop above. I don't understand why we would need to delete the files, but it might be ok.

Comment (divega) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestStore.cs]: One possible theory: we had some test that failed to recreate the database because the files were still there after the drop had executed, and deleting the files seemed to help. I wonder if DROP DATABASE is just a fire and forget operation. Maybe it would be more robust to just poll for the files before we return, yet it might not matter too much.

@anpete it seems that you added this code in https://github.com/natemcmaster/EntityFramework/commit/6eec03d4103a45f35122eb8907771f07a9195aae. Do you remember why?

File: test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlConnectionStringBuilderExtensions.cs [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+        {
+            var config = _sqlTestConfiguration.Value;
+
+            int connectTimeout;
+            if (!int.TryParse(config["ConnectTimeout"], out connectTimeout))
+            {
+                connectTimeout = DefaultConnectTimeout;
+            }
+
+            builder.ConnectTimeout = connectTimeout;
```
Comment (divega) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlConnectionStringBuilderExtensions.cs]: Should the overwrite flag apply to this too?

File: test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlConnectionStringBuilderExtensions.cs [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+
+            bool integratedSecurity;
+            if (!bool.TryParse(config["IntegratedSecurity"], out integratedSecurity))
+            {
+                integratedSecurity = DefaultIntegratedSecurity;
+            }
+
+            builder.IntegratedSecurity = integratedSecurity;
```
Comment (divega) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlConnectionStringBuilderExtensions.cs]: Overwrite too?

File: test/EntityFramework.SqlServer.FunctionalTests/MonsterFixupSqlServerTest.cs [hunk 17]
```diff
@@ -46,14 +46,14 @@ protected override DbContextOptions CreateOptions(string databaseName)
...
-                ConnectTimeout = 30
-            }.ConnectionString;
+            var connStrBuilder = new SqlConnectionStringBuilder
+                {
+                    MultipleActiveResultSets = true,
+                    InitialCatalog = name
+                };
+            connStrBuilder.LoadEnvConfiguration();
+
+            return connStrBuilder.ConnectionString;
```
Comment (divega) [inline test/EntityFramework.SqlServer.FunctionalTests/MonsterFixupSqlServerTest.cs]: Nit: it wouldn't bother me if you just chained it all inline when you can :smile:

File: test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlConnectionStringBuilderExtensions.cs (hunk 15 above)
Comment (divega) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlConnectionStringBuilderExtensions.cs]: Moreover, do you ever not overwrite? Maybe you don't need the flag.

File: test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlConnectionStringBuilderExtensions.cs [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+
+namespace Microsoft.Data.Entity.SqlServer.FunctionalTests
+{
+    public static class SqlConnectionStringBuilderExtensions
+    {
+        private static string DefaultDataSource = @"(localdb)\MSSQLLocalDB";
+        private static int DefaultConnectTimeout = 30;
+        private static bool DefaultIntegratedSecurity = true;
+
+        public static SqlConnectionStringBuilder LoadEnvConfiguration(this SqlConnectionStringBuilder builder, bool overwrite = true)
```
Comment (divega) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlConnectionStringBuilderExtensions.cs]: Nit: maybe just ApplyConfiguration?

**Comment by divega** [comment]
LGTM with some minor feedback. Should probably get some other eyes on it.

Just one last general question: do you think that this could be helpful for testing other database providers? It seems to me that with a few small tweaks the extension method could be made to work with any DbConnectionStringBuilder.

File: test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlConnectionStringBuilderExtensions.cs (hunk 15 above)
Comment (natemcmaster) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlConnectionStringBuilderExtensions.cs]: Yeah, I guess the overwrite flag isn't necessary. I'll remove it.

**Comment by natemcmaster** [comment]
Certainly it could be generalized, but we don't need something like this for SQLite (yet) so I don't think it's worth doing yet.

File: test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlConnectionStringBuilderExtensions.cs (hunk 13 above)
Comment (natemcmaster) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlConnectionStringBuilderExtensions.cs]: Nope. Moving to static constructor.

**Comment by divega** [comment]
ok, I guess @ErikEJ or @roji can send a PR if they want it generalized :smile:

File: test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestStore.cs (hunk 14 above)
Comment (divega) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestStore.cs]: Actually I was wrong on who added this code. @AndriySvyryd, do you remember why we needed to delete the files?

File: test/EntityFramework.SqlServer.FunctionalTests/MigrationsSqlServerFixture.cs [hunk 19]
```diff
@@ -16,13 +16,13 @@ protected override void ConfigureServices(IServiceCollection services)
 
         protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
         {
-            optionsBuilder.UseSqlServer("" +
-                new SqlConnectionStringBuilder
+            var connStrBuilder = new SqlConnectionStringBuilder
```
Comment (anpete) [inline test/EntityFramework.SqlServer.FunctionalTests/MigrationsSqlServerFixture.cs]: Don't abbreviate.

**Comment by divega** [comment]
@natemcmaster LGTM, but I will let @anpete do another pass and give you the shipit.

File: test/EntityFramework.SqlServer.FunctionalTests/MigrationsSqlServerFixture.cs (hunk 19 above)
Comment (anpete) [inline test/EntityFramework.SqlServer.FunctionalTests/MigrationsSqlServerFixture.cs]: Can we unify these (and the two below) to use the standard fixture pattern?

File: test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestStore.cs (hunk 14 above)
Comment (AndriySvyryd) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestStore.cs]: @divega Your theory is correct. Dropping the database didn't delete the files right away so the following tests would fail.

**Comment by natemcmaster** [comment]
@anpete https://github.com/natemcmaster/EntityFramework/commit/6c607b03b95079514b946a346d7ea44db31a5bdb using the fixture pattern.

File: test/EntityFramework.Sqlite.FunctionalTests/MonsterFixupSqliteTest.cs [hunk 20]
```diff
@@ -9,19 +9,46 @@
 using Microsoft.Data.Entity.FunctionalTests;
 using Microsoft.Data.Entity.FunctionalTests.TestModels;
 using Microsoft.Data.Entity.Infrastructure;
-using Microsoft.Data.Sqlite;
 using Microsoft.Framework.DependencyInjection;
 
 namespace Microsoft.Data.Entity.Sqlite.FunctionalTests
 {
-    public class MonsterFixupSqliteTest : MonsterFixupTestBase
+    public class MonsterFixupSqliteTest : MonsterFixupTestBase<MonsterFixupSqliteFixture>
```
Comment (anpete) [inline test/EntityFramework.Sqlite.FunctionalTests/MonsterFixupSqliteTest.cs]: Use the SqliteTestStore

File: test/EntityFramework.SqlServer.FunctionalTests/NavigationTest.cs [hunk 21]
```diff
@@ -87,10 +87,35 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
...
-            optionsBuilder.UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=StateManagerBug;Trusted_Connection=True;MultipleActiveResultSets=true");
+            var connStrBuilder = new SqlConnectionStringBuilder
+                {
+                    InitialCatalog = "StateManagerBug",
+                    MultipleActiveResultSets = true,
+                    ["Trusted_Connection"] = true
+                };
+            connStrBuilder.ApplyConfiguration();
+
+            var optionsBuilder = new DbContextOptionsBuilder();
```
Comment (anpete) [inline test/EntityFramework.SqlServer.FunctionalTests/NavigationTest.cs]: Build options in ctor and cache?

Comment (anpete) [inline test/EntityFramework.SqlServer.FunctionalTests/NavigationTest.cs]: Use test store?

File: test/EntityFramework.SqlServer.FunctionalTests/MonsterFixupSqlServerTest.cs [hunk 22]
```diff
@@ -14,14 +14,42 @@
...
     {
         private static readonly HashSet<string> _createdDatabases = new HashSet<string>();
 
         private static readonly ConcurrentDictionary<string, AsyncLock> _creationLocks
             = new ConcurrentDictionary<string, AsyncLock>();
 
-        protected override IServiceProvider CreateServiceProvider(bool throwingStateManager = false)
+        public override async Task CreateAndSeedDatabase(Func<MonsterContext> createContext)
+        {
+            using (var context = createContext())
```
Comment (anpete) [inline test/EntityFramework.SqlServer.FunctionalTests/MonsterFixupSqlServerTest.cs]: Use test store.

File: test/EntityFramework.SqlServer.FunctionalTests/MonsterFixupSqlServerTest.cs [hunk 23]
```diff
@@ -36,52 +64,79 @@ protected override IServiceProvider CreateServiceProvider(bool throwingStateMana
             return serviceCollection.BuildServiceProvider();
         }
 
-        protected override DbContextOptions CreateOptions(string databaseName)
+        private DbContextOptions CreateOptions(string databaseName)
         {
+            var connectionStringBuilder = new SqlConnectionStringBuilder
```
Comment (anpete) [inline test/EntityFramework.SqlServer.FunctionalTests/MonsterFixupSqlServerTest.cs]: Build once and cache?

File: test/EntityFramework.SqlServer.FunctionalTests/MigrationsSqlServerFixture.cs [hunk 24]
```diff
@@ -9,20 +10,29 @@ namespace Microsoft.Data.Entity.SqlServer.FunctionalTests
...
+                .ServiceCollection()
+                .BuildServiceProvider();
         }
 
-        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
+        public override MigrationsContext CreateContext()
         {
-            optionsBuilder.UseSqlServer("" +
-                new SqlConnectionStringBuilder
+            var connectionStringBuilder = new SqlConnectionStringBuilder
```
Comment (anpete) [inline test/EntityFramework.SqlServer.FunctionalTests/MigrationsSqlServerFixture.cs]: Use test store?

File: test/EntityFramework.SqlServer.FunctionalTests/MigrationsSqlServerFixture.cs [hunk 25]
```diff
@@ -9,20 +10,29 @@ namespace Microsoft.Data.Entity.SqlServer.FunctionalTests
...
                 {
-                    DataSource = @"(localdb)\MSSQLLocalDB",
-                    InitialCatalog = nameof(MigrationsSqlServerTest),
-                    IntegratedSecurity = true
-                });
+                    InitialCatalog = nameof(MigrationsSqlServerTest)
+                };
+            connectionStringBuilder.ApplyConfiguration();
+
+            var optionsBuilder = new DbContextOptionsBuilder();
```
Comment (anpete) [inline test/EntityFramework.SqlServer.FunctionalTests/MigrationsSqlServerFixture.cs]: Build once and cache?

File: test/EntityFramework.Core.FunctionalTests/MonsterFixupTestBase.cs [hunk 26]
```diff
@@ -771,29 +738,27 @@ private void Can_build_monster_model_and_seed_data_using_navigations_with_deferr
...
-            await Composite_fixup_happens_when_FKs_change_test(p => CreateChangedOnlyMonsterContext(p), ChangedOnlyDatabaseName, useDetectChanges: false);
+            await Composite_fixup_happens_when_FKs_change_test(() => Fixture.CreateChangedOnlyMonsterContext(), useDetectChanges: false);
         }
 
         private async Task Composite_fixup_happens_when_FKs_change_test(
-            Func<IServiceProvider, MonsterContext> createContext, string databaseName, bool useDetectChanges)
+            Func<MonsterContext> createContext, bool useDetectChanges)
         {
-            var serviceProvider = CreateServiceProvider();
+            await Fixture.CreateAndSeedDatabase(createContext);
```
Comment (anpete) [inline test/EntityFramework.Core.FunctionalTests/MonsterFixupTestBase.cs]: This shouldn't need to be called in each test. That's what fixtures are for. It looks like there might be multiple fixtures here and so we may need to split out each set of tests so the fixtures can vary.

File: test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestStore.cs (hunk 14 above)
Comment (divega) [inline test/EntityFramework.SqlServer.FunctionalTests/Utilities/SqlServerTestStore.cs]: Thanks. I will file a test bug on this to follow up when we have time.

File: test/EntityFramework.Core.FunctionalTests/MonsterFixupTestBase.cs (hunk 26 above)
Comment (natemcmaster) [inline test/EntityFramework.Core.FunctionalTests/MonsterFixupTestBase.cs]: I'm going to backout changes to MonsterFixup and move to a separate PR. This is expanding beyond the scope of the original PR.

Comment (natemcmaster) [inline test/EntityFramework.Core.FunctionalTests/MonsterFixupTestBase.cs]: Also, the CreateAndSeedDatabase function is not used in every test.

**Comment by natemcmaster** [comment]
Updated. Added #3007 to track monsterfixup tests

**Comment by divega** [comment]
:shipit:
